src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -99,6 +100,13 @@ pub trait StorageController: Debug + Send {
         updates: &mut BTreeMap<GlobalId, ChangeBatch<Self::Timestamp>>,
     );
 
+    /// Send a request to obtain "linearized" timestamps for the given sources.
+    async fn linearize_sources(
+        &mut self,
+        peek_id: Uuid,
+        source_ids: Vec<GlobalId>,
+    ) -> Result<(), anyhow::Error>;
```

Comment:
@frankmcsherry @benesch I was wondering, why do we go for this pattern where these async methods fire a command to storage and then the caller needs to observe the response stream to figure out what happened?

It sounds to me that this function interface would be much better for users interacting with STORAGE:

```rust
async fn linearize_sources(&mut self, source_ids: Vec<GlobalId>) -> Result<Self::Timestamp, anyhow::Error>;
```

i.e you get back a `Future` that will resolve to your result and there is no need to specify a `peek_id`. It's the responsibility of the storage controller implementation to figure out a way to match a request to a response and you don't have to dissect the `recv` stream to figure out what happened. 

Do we lose something if we design it that way?

Reply:
I'm not sure! We potentially lose some visibility of the command and response flow (currently `trace!`d output). I find that very helpful for the compute commands (e.g. why not do the same for `Peek`?) but I could also imagine that we could keep that behaviour and wrap it up differently if we want.

There is a more complicated multi-way `select!` to write I think, if we have 1k outstanding requests of this form, that a channel addresses nicely.

My personal current feelings are that `async` is a grim antipattern and the less we do of it the better (for reasons of cancelation safety), but that may just be due to being burned by it recently. Putting things in a queue is much more clearly not-wrong, for me at least.

Reply:
@petrosagg In the work I am doing that follows this, `linearize_sources` is async, but all it does it `spawn` a tokio task  that will eventually send a response back on the channel back to the coordinator. In that sense, this method doesn't _really_ need to be async, and we can simplify it later if need be. If we did all the work inline in this `async fn`, then we would hang the coordinator waiting for sources to catch up to the linearized offsets